core/src/protocols_handler/mod.rs

```diff
@@ -140,22 +140,22 @@ pub trait ProtocolsHandler {
     /// therefore no more inbound substreams will be produced.
     fn inject_inbound_closed(&mut self);
 
-    /// Returns whether the connection should be kept alive.
+    /// Returns until when the connection should be kept alive.
     ///
-    /// If returns `false`, that indicates that this connection is not important and the user may
-    /// invoke `shutdown()` if they think that they will no longer need the connection in the
-    /// future.
+    /// If returns `Some`, that indicates that this connection may invoke `shutdown()` after the
+    /// returned `Instant` has elapsed if they think that they will no longer need the connection
+    /// in the future. Returning `None` is equivalent to "infinite".
```

Comment:
Instead of `Option<Instant>` I would prefer a custom enum type.  That returning `None` from `connection_keep_alive` is defined to mean forever reads a little strange. How about:

```rust
pub enum KeepAlive {
    Until(Instant),
    Forever
}
```